HW-02 Model 적용하기

onemask opened the pull request:
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/132
에서 받았던 피드백 위주로 작업하였고 
석준님이 말씀하셨던 comment 반영하였습니다.

onemask committed:
Edit > UpbitRepository는 UpbitDataSource를 implement하여 구현

onemask committed:
Edit > UpbitDatasource > UpbitApi로 변경,

onemask committed:
Edit > model data의 prefix다 Upbit로 통일

onemask committed:
Edit > 기존 modelType prefix로 Upbit추가

onemask committed:
Delete  > 기존 repository 파일명 변경

onemask committed:
Edit > MainActivitym,Fragment에서 repository로 API로 호출하게끔 변경 & lazy문을 이용하여 람다안의 객체 생성 적용

onemask committed:
Edit > 클래스명 UpbitDatasource  > UpbitDataSource로 변경

preludezdev commented on onemask/ArchitectureStudy/app/src/main/java/com/android/studyfork/data/repository/UpbitDatasource.kt:
```diff
@@ -0,0 +1,13 @@
+interface UpbitDataSource {
+    fun getMarketAll(): Single<Map<String, List<UpbitMarketResponse>>>
+    fun getTikcers(margket: String): Single<List<UpbitTickerResponse>>
+}
```
getTikcers, margket, 인터페이스 implement 하는 함수들까지 오타 수정해주시고 파일명에도 파스칼케이스 적용해주세요

onemask replied:
수정하였습니다. 
꼼꼼히 봐주셔서 감사합니다. 
a3e385c

preludezdev replied:
getTickers() 에 K 가 대문자로 되어있습니다...ㅠㅠ

onemask replied:
다시금  꼼꼼히 봐주셔서 감사합니다..ㅠㅠ
수정하였습니다. 

051568c, a51aa8a

preludezdev commented on onemask/ArchitectureStudy/app/src/main/java/com/android/studyfork/network/api/UpbitApi.kt:
```diff
@@ -0,0 +1,17 @@
+
+    @GET("ticker")
+    fun getTikers(@Query("markets") markets: String?): Single<List<UpbitTickerResponse>>
+
```
getTikers 오타 수정해주세요

onemask replied:
수정하였습니다. 
75b6cad

preludezdev commented on onemask/ArchitectureStudy/app/src/main/java/com/android/studyfork/ui/TikcerListFragment.kt:
```diff
@@ -9,21 +9,19 @@ import androidx.fragment.app.Fragment
 import timber.log.Timber
 
 class TikcerListFragment : Fragment() {
 
```
클래스명 오타 수정해주세요

onemask replied:
수정하였습니다.
26d1ca0

preludezdev commented on onemask/ArchitectureStudy/app/src/main/java/com/android/studyfork/data/repository/UpbitRepository.kt:
```diff
@@ -1,23 +1,24 @@
+class UpbitRepository(
+    private val api: UpbitApi
+) : UpbitDataSource {
 
```
석준님이 저번 수연씨 PR에 적어준 내용보면 레포지토리 패턴은 Repository에서 RemoteDataSource에 데이터를 요청하면 RemoteDataSource에서 Retrofit을 사용해 데이터를 가져오는 형태라고 하셨는데 지금 올려주신 코드는 Repository 클래스 내에서 레트로핏 통신을 해서 데이터를 가져오고 있는 것 같아요. 
제가 이해하기로는 Repository 클래스랑 RemoteDataSource를 분리해야 되서 아래 그림처럼 구조를 고쳐야된다고 생각하는데 이부분은 확실하게 맞는지 석준님이 체크해주셨음 좋겠어요 @sjjeong 석준님 확인 부탁드릴게요...!!
![repository 예시](https://user-images.githubusercontent.com/41214420/61791538-3bc91d80-ae55-11e9-8856-6230ab2c97cc.jpg)

코드는 구글 mvp 예제 보시면 도움 될거에요!

[google mvp sample link]
https://github.com/googlesamples/android-architecture/tree/todo-mvp-kotlin/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source

onemask replied:
음 말씀해주신 내용확인 해봤는데 현재 제 코드에서 
UpbitRepository가 UpbitDataSource를 implement하고 
그 안에  생성자 파라미터로 UpbitApi를 이용하여 API 호출을 하고 있습니다. 

명명을 어떻게 지어야 할지 모르겠지만 현재 저 그림에서 
UpitDataSource가 현재 제 UpbitRepository라고 보시면 되시고
UpbitRepository까 저의 UpitDataSource락고 보시면 될거 같아요. 

Repository패턴에 대하여는 잘 모르겠지만 
repositorry를 이용하여 api를 호출하고 싶어서 현재 제 코드대로 구현하였습니다.

preludezdev replied:
말씀해주신대로라면 수연님의 UpbitRepository가 UpbitDataSource는 될 수 있겠지만
UpbitDataSource는 인터페이스라서 UpbitRepository가 될 수 없을 것 같습니다. 

google MVP sample 을 보시면 Repository는 싱글톤으로 구현된 클래스로 뷰에서 데이터를 가져오고 싶을 때 해당 뷰에서 Repository.getInstance() 로 객체를 가져와 데이터에 접근하고 있습니다.

[google mvp sample - Repository 클래스]
https://github.com/googlesamples/android-architecture/blob/todo-mvp-kotlin/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepository.kt

보시고 확인 부탁드립니다

preludezdev replied:
모델 구현하는데 도움이 될만한 글을 찾아서 링크 걸어드립니다
https://thdev.tech/androiddev/2017/01/09/Android-MVP-Model-Two/
https://thdev.tech/androiddev/2017/01/29/Android-MVP-Google-Architecture-Model/

onemask replied:
[google mvp sample - Repository 클래스]
h](https://github.com/googlesamples/android-architecture/blob/todo-mvp-kotlin/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepository.kt)
알려주신 좋은 예제를 참고하여 수정하였습니다. 
싱글톤으로 접근할수 있게 동일하게 작업하였습니다. 
1180ead

onemask replied:
@preludezdev 
참고로 보내 주신 링크중에서 [Repository](https://thdev.tech/androiddev/2017/01/09/Android-MVP-Model-Two/)에서 Reposistory에 대한 구현을 
`public class ImageRepository implements ImageSource {

    private ImageRemoteDataSource imageRemoteDataSource;

    private static ImageRepository imageSampleRepository;

    public static ImageRepository getInstance() {
        if (imageSampleRepository == null) {
            imageSampleRepository = new ImageRepository();
        }

        return imageSampleRepository;
    }

    private ImageRepository() {
        imageRemoteDataSource = ImageRemoteDataSource.getInstance();
    }

    @Override
    public void getImageItems(int page, final LoadImageCallback loadImageCallback) {
        imageRemoteDataSource.getImageItems(page, loadImageCallback);
    }
}`

이런 식으로 정의하셨는데 현재 저의 Repository에서도 위와 같은 구조인데
그리고 사실상 싱글톤으로 접근할수 있게 IRepository INSTANCE를 빼 놓아도 
그 안에 apiservice가 Instance로 초기화가 되어야 해서 lazy로 구현을 해놓은 상태입니다. 

그래서 뭔가 제대로 Instance화해서 접근한 기분이 아니라 
태휘님이 말씀하셨던 방향의  Repository인지 여쭤보고 싶었습니다.

preludezdev replied:
제가 말씀드리고 싶었던 부분은 Repository 클래스와 Remote/Local DataSource 클래스가 분리된 다른 클래스로 정의되어야 한다는 것이었습니다. 위에 걸어둔 링크 내용처럼 모델의 각 클래스 정의는 다음과 같다고 합니다.

**Repository** : 로컬/서버 중 어떤 데이터를 불러올지 정의하고, 메모리 캐시를 포함한다.
**Remote Data Source** : 서버에서 데이터를 받아온다
**Local Data Source** : 로컬에서 데이터를 받아온다 

이렇게 클래스를 Repository와 Remote/Local DataSource 로 분리하는 이유는 원활한 유지보수와 재사용성을 위해서라고 들었습니다. 예를 들어 Repository와 Remote/Local DataSource가 분리되어있지 않을 때 Upbit Api와 통신하는 부분에 코드 변경이 필요하게 되면 전체 Repository 클래스를 변경해야 되서 복잡하지만, 만약 저 셋이 잘 분리되어있다면 Repository 는 그대로 두고 Remote DataSource 클래스만 변경하면 되기 때문에 코드 개선이 용이해집니다. 그래서 기본적으로 Model은 아래 그림처럼 셋이 분리되어있는 구조여야 한다고 들었습니다.

![image](https://user-images.githubusercontent.com/41214420/61970783-73cc8e00-b018-11e9-8bfa-113bdd1483fd.png)

위 예제의 ImageRepository 클래스의 경우에도 Repository 클래스 내부에 Image**Remote**DataSource 객체를 프로퍼티로 가지고 해당 프로퍼티를 이용해 Image**Remote**DataSource의 함수를 호출해서 데이터에 접근하는 걸 볼 수 있습니다.

이 부분이 해결되면 지금처럼 레포지토리에 직접 api를 넣을 필요 없이 레트로핏 통신을 하는 Remote DataSource 클래스 안에서 api를 관리하면 되니 말씀해주신 부분도 조금 더 개선이 가능하시지 않을까 생각합니다.

preludezdev reviewed (commented):
부족하지만 피드백 남겼습니다!

preludezdev commented:
데이터를 받아서 아이템뷰에 bind해서 뿌려주는 부분에서 데이터를 가공하는 부분이 extension 이용해서 구현하셨던데 어댑터에 있으면 안된다고 했습니다. 확인 부탁드릴게요.
그리고 아시겠지만 데이터 가공하는 로직들도 개선해주세요. (데이터 크기별로 M,k 붙이기, 1000단위로 ',' 붙이기 등)

onemask committed:
HW-2 UpbitDataSource > getTickers 오타 수정,market 오타 수정

onemask committed:
HW-2 UpbitApi > getTickers 오타 수정

onemask committed:
HW-2 TkcerListFragment 클래스명 수정

onemask commented:
> * CoinItemAdapter 클래스에 데이터 받아서 가공하는 부분을 extension 이용해서 구현하셨던데 가공로직을 뷰에서 하게 되면 작업이 무거워진다고 하네요. (새로운 모델클래스 정의 필요) 확인 부탁드릴게요.
> * 데이터 가공하는 작업도 개선해주세요. (데이터 크기별로 M,k 붙이기, 1000단위로 ',' 붙이기 등)
> * 커밋명은 HW-n 으로 시작해주세요

onemask commented:
> > * CoinItemAdapter 클래스에 데이터 받아서 가공하는 부분을 extension 이용해서 구현하셨던데 가공로직을 뷰에서 하게 되면 작업이 무거워진다고 하네요. (새로운 모델클래스 정의 필요) 확인 부탁드릴게요.
> > * 데이터 가공하는 작업도 개선해주세요. (데이터 크기별로 M,k 붙이기, 1000단위로 ',' 붙이기 등)
> > * 커밋명은 HW-n 으로 시작해주세요

onemask commented:
comment를 다시 날리려고 했는데 해당 pr을 close하였다가 다시 reopen하게 되었습니다

onemask commented:
@preludezdev 

- [ ] CoinItemAdapter 클래스에 데이터 받아서 가공하는 부분을 extension 이용해서 구현하셨던데 가공로직을 뷰에서 하게 되면 작업이 무거워진다고 하네요. (새로운 모델클래스 정의 필요) 확인 부탁드릴게요.
이라는 말씀은 가공로직을 따로 하는 클래스를 말씀하시는건가요?

preludezdev commented:
> @preludezdev
> 
> * [ ]  CoinItemAdapter 클래스에 데이터 받아서 가공하는 부분을 extension 이용해서 구현하셨던데 가공로직을 뷰에서 하게 되면 작업이 무거워진다고 하네요. (새로운 모델클래스 정의 필요) 확인 부탁드릴게요.
>   이라는 말씀은 가공로직을 따로 하는 클래스를 말씀하시는건가요?

그것보단 데이터 가공 시점의 문제인 것 같습니다. 석준님이 뷰에서는 받은 데이터를 그저 보여주기만 하지 데이터 가공같은 무거운 작업들은 하면 안된다 하셔서 저는 데이터를 받는 시점인 Fragment에서 데이터를 가공해서 어댑터에 넘겨주었습니다.

onemask committed:
HW-2 UpbitRepository 싱글톤으로 뺄 수 있게 Instance 작업

onemask committed:
HW-2 getTickers 오타 수정

onemask committed:
HW-2 Tickers 대문자 수정

onemask commented:
@preludezdev 
말씀하신 데이터 가공이라는것이 뷰에 거래대금이난 전일대비에 따라서 색 변환나 데이터 포맷팅에 대하여 
말씀해주시는건가요? 저는 해당 작업을 Util의 TextView로 작업하였는데..
그 Extension으로 작업하는것도 무거운 작업일까요..? 
@sjjeong-dino

sjjeong commented:
아직 리뷰가 끝나지 않은 것 같아서 Mentor Needed는 제거 했습니다.

sjjeong commented:
> @preludezdev
> 말씀하신 데이터 가공이라는것이 뷰에 거래대금이난 전일대비에 따라서 색 변환나 데이터 포맷팅에 대하여
> 말씀해주시는건가요? 저는 해당 작업을 Util의 TextView로 작업하였는데..
> 그 Extension으로 작업하는것도 무거운 작업일까요..?
> @sjjeong-dino

결국 extension에서 하더라도 onBindViewHolder 시점에서 하는 것이기 때문에 동일합니다.
제가 말씀드렸던 것은 onBindViewHolder시점이 아닌 데이터를 Adapter에 set하기 전에 가공하라는 것이였습니다.

sjjeong commented:
저는 @sjjeong 로 해주시면 됩니다.

onemask committed:
HW -2 사용하지 않는 TextView Extension 삭제
      기존 TextView Extension에서 Double형으로 꾜체

onemask committed:
HW -2 CoinItemAdapter item 모델 교체,
      Data를 Foramatin할 수 있는 Model Type 추가
      TickerListFragment에서 map으로 데이터를 필터링하여 Adapter에       넣어줌

onemask commented:
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pulls#issuecomment-515519675
에 대하여 수정하였습니다. 
b8742b1 , e17eebe

sjjeong commented:
수연님께서 3기에서 4기로 이동하면서 branch도 class03/onemask에서 class04/onemask로 이동했기 때문에 해당 pr은 close하겠습니다.
